### PR TITLE
chore: update mini-message and inventory projects

### LINKED ISSUE #27927: Update Projects: Mini message board && inventory application
## Lesson overview

Update projects to align with Node Revamp

## Lesson outline

The following items define the scope of the lesson. The lesson outline is flexible; you can improve, expand, and omit items whilst writing lesson content (make sure the pull request description mentions that).

- asdf 

## Acceptance criteria

If the requirements here are not met, the work effort is not complete.

- [ ] Lesson conforms to the outline above (unless outline has been modified).
- [ ] Lesson follows our [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md).

### Get the lesson template

Download the lesson template using the following command (replace `lesson_name` with the actual lesson name):

```bash
curl -o <lesson_name>.md https://raw.githubusercontent.com/TheOdinProject/curriculum/main/templates/lesson-template.md 
```

### PR DESCRIPTION
**Pull request by 01zulfi**
## Because

https://github.com/TheOdinProject/curriculum/issues/27927

## This PR

https://github.com/TheOdinProject/curriculum/issues/27927

## Issue

Closes #27927 

## Pull Request Requirements

-   [ ] I have thoroughly read and understand [The Odin Project curriculum contributing guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md)
-   [ ] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [ ] The `Because` section summarizes the reason for this PR
-   [ ] The `This PR` section has a bullet point list describing the changes in this PR
-   [ ] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [ ] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Review by MaoShizhong — COMMENTED** [review]
How will URLs be handled for links to other lessons in the course, specifically the ones that are brand new files in the `express/` directory (and not the ones that are updating existing lesson files), e.g. the new deployment lesson? Will they retain the same URL as the lessons they're replacing?

File: nodeJS/express/project_inventory_application.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+1. Before you begin, take a moment to write down all of the database tables and its fields you'll need, as well as the relations between them. For example:
+    1. In a game management app, there can be a game, genre, and developer entity. A game can have one or multiple developers and genres. Similarly a developer can develop multiple games.
+    1. In a pokemon management app, there can be a pokemon, trainer and a type entity. Each pokemon must be contained in a type. While a trainer can have multiple pokemons.
+
+    Any sufficient inventory app will have relations and constraints against its entities. Figure out these database particulars for your inventory app.
+1. Set up the routes and controllers you're going to need.
+1. Create all of the 'READ' views (i.e. view category, and view item)
+1. Create all the forms and build out the controllers you need for the create and update actions.
+1. Figure out the delete functionality. What happens if you delete a category with items in it? Should it delete all the items as well? Or should it just remove the category from the items? This specific behavior will depend on your app's requirements.
```
Comment (MaoShizhong) [inline nodeJS/express/project_inventory_application.md:22]: Possibly makes it more "you decide with what makes sense to you" than "these are your only two options - pick one" to me

```suggestion
1. Figure out the delete functionality. What happens if you try to delete a category with items in it? Should it delete all the items as well? Should it just remove the category from the items? Or something else? This specific behavior will depend on your app's requirements.
```

File: nodeJS/express/project_inventory_application.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+<div class="lesson-content__panel" markdown="1">
+
+1. Set up an Express project and a new PostgreSQL database.
+1. Before you begin, take a moment to write down all of the database tables and its fields you'll need, as well as the relations between them. For example:
+    1. In a game management app, there can be a game, genre, and developer entity. A game can have one or multiple developers and genres. Similarly a developer can develop multiple games.
+    1. In a pokemon management app, there can be a pokemon, trainer and a type entity. Each pokemon must be contained in a type. While a trainer can have multiple pokemons.
+
+    Any sufficient inventory app will have relations and constraints against its entities. Figure out these database particulars for your inventory app.
+1. Set up the routes and controllers you're going to need.
+1. Create all of the 'READ' views (i.e. view category, and view item)
```
Comment (MaoShizhong) [inline nodeJS/express/project_inventory_application.md:20]:
```suggestion
1. Create all of the 'READ' views (i.e. view category, and view item).
```

File: nodeJS/express/project_mini_message_board.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
+### Introduction
+
+At this point you should know enough to use Express to make some fun interactive web apps! We're going to create a message board.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Set up a basic Express app by installing Express and EJS. Set up a basic index route and run your server. Create required folders and files as discussed in the previeous lessons.
```
Comment (MaoShizhong) [inline nodeJS/express/project_mini_message_board.md:9]:
```suggestion
1. Set up a basic Express app by installing Express and EJS. Set up a basic index route and run your server. Create the required folders and files as discussed in the previous lessons.
```

File: nodeJS/express/project_mini_message_board.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+1. We are going to have 2 routes, the index (`"/"`) and a new-message form (`"/new"`).
```
Comment (MaoShizhong) [inline nodeJS/express/project_mini_message_board.md:10]: Nit
```suggestion
1. We are going to have 2 routes, the index (`"/"`) and a "new message" form (`"/new"`).
```

File: nodeJS/express/project_mini_message_board.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+     },
+     {
+       text: "Hello World!",
+       user: "Charles",
+       added: new Date()
+     }
+   ];
+   ```
+
+1. Next, in your index template (in the `"views"` folder) loop through the messages array using whichever templating language you selected and for each one, display the user, text and the date the message was added. Don't forget to make your messages available to your template by including it in the res.render 'locals' object (e.g. `res.render('index', { title: "Mini Messageboard", messages: messages })`).
```
Comment (MaoShizhong) [inline nodeJS/express/project_mini_message_board.md:28]:
```suggestion
1. Next, in your index template (in the `"views"` folder) loop through the messages array and for each one, display the user, text and the date the message was added. Don't forget to make your messages available to your template by including it in the `res.render` 'locals' object (e.g. `res.render("index", { title: "Mini Messageboard", messages: messages })`).
```

File: nodeJS/express/project_mini_message_board.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+   app.use(express.urlencoded({ extended: true }));
+   ```
+
+1. In your `router.post()` take the contents of the form submission and push them into the messages array as an object that looks something like this:
+
+   ```javascript
+   messages.push({ text: messageText, user: messageUser, added: new Date() });
+   ```
+
+1. At the end of the `router.post()` function use `res.redirect('/')` to send users back to the index page after submitting a new message.
```
Comment (MaoShizhong) [inline nodeJS/express/project_mini_message_board.md:50]:
```suggestion
1. At the end of the `router.post()` function use `res.redirect("/")` to send users back to the index page after submitting a new message.
```

File: nodeJS/express/project_mini_message_board.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+1. At this point, you should be able to visit `/new` (it might be a good idea to add a link to that route on your index page), fill out the form, submit it and then see it show up on the index page!
+1. Add a "open" button against every message to open a new page with the message details.
```
Comment (MaoShizhong) [inline nodeJS/express/project_mini_message_board.md:52]:
```suggestion
1. Add an "open" button next to every message to open a new page with the message details.
```

File: nodeJS/final_project/project_odin_book.md
Comment (MaoShizhong) [inline nodeJS/final_project/project_odin_book.md:1]: While we're at it, there are two mentions of `passportJS` on line 7 here where the first mention should be changed to `passport.js, and the second mention either the same or changed to simply `Passport`.

**Review by MaoShizhong — CHANGES_REQUESTED** [review]
:fire:

**Comment by 01zulfi** [comment]
> How will URLs be handled for links to other lessons in the course, specifically the ones that are brand new files in the express/ directory (and not the ones that are updating existing lesson files), e.g. the new deployment lesson which is linked in line 34 of the blog project? Will they retain the same URL as the lessons they're replacing?

I'm not sure if they'll retain the same URL. We can update those URLs when the revamp goes live though. In case, we do miss some, we can rely on the community to submit issues/PRs.

**Comment by MaoShizhong** [comment]
> > How will URLs be handled for links to other lessons in the course, specifically the ones that are brand new files in the express/ directory (and not the ones that are updating existing lesson files), e.g. the new deployment lesson which is linked in line 34 of the blog project? Will they retain the same URL as the lessons they're replacing?
> 
> I'm not sure if they'll retain the same URL. We can update those URLs when the revamp goes live though. In case, we do miss some, we can rely on the community to submit issues/PRs.

In that case, I'll make a note of all relevant links in the new lessons and once they go live, do a check and PR href changes as necessary.

**Review by MaoShizhong — APPROVED** [review]
:rocket:

**Review by MaoShizhong — CHANGES_REQUESTED** [review]
Oops, nearly missed this one!

OdinBook line 46 currently still references the Inventory Application project for Cloudinary, but Cloudinary will be moved to the File Uploader project instead.

We should edit that line to:

```md
1. Make posts also allow images (either just via a URL or by uploading one). If you did the extra credit from [File Uploader project](https://www.theodinproject.com/lessons/nodejs-file-uploader), then you may recall [Cloudinary](https://cloudinary.com/documentation/node_integration) or [Supabase storage](https://supabase.com/docs/guides/storage) being good options for hosting user-uploaded images. The URLs they provide you can then be stored in your database instead of the raw image binary data.
```

**Review by MaoShizhong — APPROVED** [review]

**Pull request merged**